src/ophyd_async/core/_signal_backend.py

```diff
@@ -1,97 +1,163 @@
-        else:
-            if not isinstance(_choices, tuple) or not all(
-                isinstance(c, str) for c in _choices
-            ):
-                raise TypeError(
-                    "Choices must be a str or a tuple of str, " f"not {type(_choices)}."
-                )
-            if len(set(_choices)) != len(_choices):
-                raise TypeError("Duplicate elements in runtime enum choices.")
-
-        class _RuntimeSubsetEnum(cls):
-            choices = _choices
-
-        return _RuntimeSubsetEnum
-
-
-class RuntimeSubsetEnum(metaclass=_RuntimeSubsetEnumMeta):
-    choices: ClassVar[tuple[str, ...]]
-
-    def __init__(self):
-        raise RuntimeError("SubsetEnum cannot be instantiated")
-
-
-if TYPE_CHECKING:
-    SubsetEnum = Literal
-else:
-    SubsetEnum = RuntimeSubsetEnum
+_primitive_dtype: dict[type[Primitive], Dtype] = {
+    bool: "boolean",
+    int: "integer",
+    float: "number",
+    str: "string",
+}
+
+
+class SignalMetadata(TypedDict, total=False):
+    limits: Limits
+    choices: list[str]
+    precision: int
+    units: str
+
+
+def _datakey_dtype(datatype: type[SignalDatatype]) -> Dtype:
+    if (
+        datatype is np.ndarray
+        or get_origin(datatype) in (Sequence, np.ndarray)
+        or issubclass(datatype, Table)
+    ):
+        return "array"
+    elif issubclass(datatype, StrictEnum):
+        return "string"
+    elif issubclass(datatype, Primitive):
+        return _primitive_dtype[datatype]
+    else:
+        raise TypeError(f"Can't make dtype for {datatype}")
+
+
+def _datakey_dtype_numpy(
+    datatype: type[SignalDatatypeT], value: SignalDatatypeT
+) -> np.dtype:
+    if isinstance(value, np.ndarray):
+        # The value already has a dtype, use that
+        return value.dtype
+    elif (
+        get_origin(datatype) == Sequence
+        or datatype is str
+        or issubclass(datatype, StrictEnum)
+    ):
+        # TODO: use np.dtypes.StringDType when we can use in structured arrays
+        # https://github.com/numpy/numpy/issues/25693
+        return np.dtype("S40")
+    elif isinstance(value, Table):
+        return value.numpy_dtype()
+    elif issubclass(datatype, Primitive):
+        return np.dtype(datatype)
+    else:
+        raise TypeError(f"Can't make dtype_numpy for {datatype}")
+
+
+def _datakey_shape(value: SignalDatatype) -> list[int]:
+    if type(value) in _primitive_dtype or isinstance(value, StrictEnum):
+        return []
+    elif isinstance(value, np.ndarray):
+        return list(value.shape)
+    elif isinstance(value, Sequence | Table):
+        return [len(value)]
+    else:
+        raise TypeError(f"Can't make shape for {value}")
+
+
+def make_datakey(
+    datatype: type[SignalDatatypeT],
+    value: SignalDatatypeT,
+    source: str,
+    metadata: SignalMetadata,
+) -> DataKey:
+    dtn = _datakey_dtype_numpy(datatype, value)
+    return DataKey(
+        dtype=_datakey_dtype(datatype),
+        shape=_datakey_shape(value),
```

Comment:
We are saying that `dtype_numpy` expresses a single row, so shape should be `(num_rows,)`. We did this because the columns can all be of different types, so can't be expressed as a 2D structure.